### PR TITLE
feat(assessors): add lint suppression density assessor

### LINKED ISSUE #510: [FEATURE] Add lint suppression density check to detect backpressure erosion
## Problem Statement

A repo can have a comprehensive lint config and still render it meaningless through suppression directives. If large portions of the codebase use //nolint, # noqa, or // eslint-disable, the enforcement signal degrades — lint passes, but not because the code is clean. agentready currently cannot distinguish a genuinely lint-clean codebase from one that's heavily suppressed. An agent will treat lint output as a reliable quality signal when it isn't.

## Proposed Solution

Scan source files for suppression directives by language, normalize against codebase size, and score based on density:

Go: //nolint
Python: # noqa, # type: ignore, # pylint: disable
JavaScript/TypeScript: // eslint-disable, // @ts-ignore
Ruby: # rubocop:disable
Metric: suppressions per 1,000 lines of source code. Suggested thresholds: pass at ≤5/1k, warning at 6–15/1k, fail above 15/1k. Thresholds should be validated against real-world distribution across assessed repos and made configurable via .agentready-config.yaml.

## Alternatives Considered

Binary check: does any suppression exist? Too blunt — some suppressions are legitimate (e.g., suppressing a false positive in generated code). Density normalizes for codebase size and allows for legitimate use while flagging systemic abuse.

## Use Cases

Describe specific use cases where this feature would be valuable.

1. A repo passes lint in CI but has 200 //nolint directives across 5k lines — an agent working in it can't trust lint output as a quality signal.
2. A team reducing suppression density over time sees their agentready score improve as enforcement becomes more meaningful.
3. Assessment report identifies which files have the highest suppression concentration, giving teams a specific remediation target.

## Additional Context

Pairs with lint config coverage (see related feature request). A repo that scores well on coverage but high on suppression density has enforcement in name only. The two checks together give a complete picture of whether lint backpressure is real.

## Acceptance Criteria

- Suppression directives detected for Go, Python, JS/TS (at minimum)
- Count normalized per 1,000 lines of source code (excluding test files optionally)
- Score thresholds are configurable via .agentready-config.yaml
- Assessment evidence reports total suppression count, LOC, and density ratio
- Assessment report highlights files with highest suppression concentration

## Priority

How important is this feature to you?

- [ X] Critical (blocking my work)
- [ ] High (significantly improves workflow)
- [ ] Medium (nice to have)
- [ ] Low (minor enhancement)

### PR DESCRIPTION
**Pull request by msu8**
## Description

Add `LintSuppressionAssessor` to detect overuse of lint suppression directives
across the codebase. A repo can have a comprehensive lint config and still
render it meaningless through blanket `//nolint`, `# noqa`, or
`// eslint-disable` usage — lint passes, but not because the code is clean.
This gives AI agents a false signal that the codebase is healthy.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update
- [x] Test coverage improvement

## Related Issues

Fixes #510

## Changes Made

- New `LintSuppressionAssessor` (`lint_suppression_density`, Tier 3, 2% weight)
- Scans source files for suppression directives across 11 languages: Go (`//nolint`), Python (`# noqa`, `# type: ignore`, `# pylint: disable`), JavaScript/TypeScript (`// eslint-disable`, `// @ts-ignore`), Ruby (`# rubocop:disable`), Java (`@SuppressWarnings`), Terraform (`# tflint-ignore:`), Shell (`# shellcheck disable=`), Dockerfile (`# hadolint ignore=`), YAML (`# yamllint disable`, `# kube-linter disable`), Markdown (`<!-- markdownlint-disable -->`)
- Normalizes count per 1,000 LOC; scores 100 at ≤5/1k, 0 at ≥15/1k, linear in between
- Thresholds and test-file exclusion configurable via `.agentready-config.yaml::assessor_options`
- Excludes `vendor/`, `node_modules/`, and other generated directories from scanning
- 47 unit tests covering all 11 languages, pass/fail/partial scoring, custom thresholds, test exclusion, excluded dirs, evidence content, and attribute metadata
- `docs/attributes.md` updated with scoring rules, language coverage, and remediation examples
- `default-weights.yaml` updated (v2.5.0, 30 total attributes)
- Consolidated into `code_quality.py` — no separate module

## Testing

- [x] Unit tests pass (`pytest`)
- [x] Integration tests pass
- [x] Manual testing performed
- [x] No new warnings or errors

Manual test: assessed `psf/black` (Python), `golangci/golangci-lint` (Go) and others

image

image

## Checklist

- [x] My code follows the project's code style
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Additional Notes
-

## Summary by CodeRabbit

- **New Features**
  - Added “Lint Suppression Density” assessment with multi-language scanning, suppression density scoring, evidence reporting, and remediation guidance.
- **Documentation**
  - Expanded the attributes overview to 30 total, updated tier counts/weights, and updated implementation status for the full assessor set.
- **Tests**
  - Added unit coverage for suppression detection, scoring thresholds (including boundary cases), test-file exclusions, and assessor registration.
- **Chores**
  - Updated default weights and rebalanced related tier weight contributions.
  - Enhanced language detection to recognize additional extensions and extensionless filenames (e.g., Dockerfile).

### CODE REVIEW

File: src/agentready/assessors/code_quality.py [hunk 1]
```diff
@@ -1925,3 +1926,313 @@ def _create_remediation(self, missing: list[str], language: str) -> Remediation:
...
+            extensions = _LANG_EXTENSIONS[lang]
+            patterns = _SUPPRESSION_PATTERNS[lang]
+            for src_file, rel in self._walk_source_files(repository.path, extensions):
+                if exclude_tests and self._is_test_file(rel, lang):
+                    continue
+                sup_count, line_count = self._count_file_suppressions(src_file, patterns)
+                total_suppressions += sup_count
+                total_lines += line_count
+                if sup_count > 0:
+                    file_stats.append((sup_count, rel))
```
[the remark this answers is not in the document]
Comment (msu8) [reply src/agentready/assessors/code_quality.py:2265]: create_pass/fail aren't implemented and are only documented in .md files, implementation is out of the scope for this PR. The assessor is using the direct constructor which is the established pattern.

File: src/agentready/assessors/code_quality.py [hunk 2]
```diff
@@ -1925,3 +1926,315 @@ def _create_remediation(self, missing: list[str], language: str) -> Remediation:
...
+        file_path: Path,
+        patterns: list[re.Pattern],
+    ) -> tuple[int, int]:
+        try:
+            text = file_path.read_text(encoding="utf-8", errors="ignore")
+        except OSError:
+            return 0, 0
+        lines = text.splitlines()
+        sup_count = sum(1 for line in lines if any(p.search(line) for p in patterns))
+        return sup_count, len(lines)
```
[the remark this answers is not in the document]
Comment (msu8) [reply src/agentready/assessors/code_quality.py]: Scope creep with barely any impact in production code.

**Comment by msu8** [comment]
Hi @jwm4, could you PTAL?

**Comment by kami619** [comment]
@jwm4 and @msu8 I am taking a look at this PR.

**Comment by kami619** [comment]
## 🤖 AI-assisted review

> **Disclosure.** This review was produced by an AI agent, not a human reviewer. Every finding below was reproduced empirically against commit `6b3581f` before being reported — but treat the analysis as a starting point for human judgment, not as a verdict. Reproduction commands are included so you can verify each claim yourself.
>
> | | |
> |---|---|
> | **Model** | Claude Opus 4.6 (`claude-opus-4-6`) |
> | **Harness** | Claude Code CLI |
> | **Skill** | `/pr-review-toolkit:review-pr` |
> | **Sub-agents** | `code-reviewer`, `pr-test-analyzer`, `silent-failure-hunter`, `type-design-analyzer`, `comment-analyzer` (5, run independently in parallel) |
> | **Scope** | `git diff main..HEAD` @ `6b3581f4f470a5f7feab345ccc98684598468656` |

Five specialist agents reviewed independently and converged on the same core defect from four different angles. Findings are deduplicated and ranked below.

---

## 🔴 Critical — merge blockers

### 1. The branch breaks CI

[`tests/unit/test_assessors_structure.py:1513`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/tests/unit/test_assessors_structure.py#L1510-L1514) still asserts `default_weight == 0.02`, but [`structure.py:1403`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/structure.py#L1400-L1405) was changed to `0.01` to fund the new attribute.

```
$ pytest tests/unit/test_assessors_structure.py -q
E   AssertionError: assert 0.01 == 0.02
1 failed, 74 passed
```

Green on `main` (75/75), red here. One-line fix.

### 2. Terraform and Dockerfile support is unreachable dead code

*Found independently by all five agents.*

[`LanguageDetector.EXTENSION_MAP`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/services/language_detector.py#L20-L59) has no `.tf` entry, and detection is suffix-only (`path.suffix`), so an extensionless `Dockerfile` can never match. `repository.languages` therefore can never contain `"Terraform"` or `"Dockerfile"` — making the pattern entries at [`code_quality.py:1961,1963`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L1955-L1965) inert, since both `is_applicable()` and the scan loop intersect against that dict.

Reproduced end-to-end on a repo containing 20 `# tflint-ignore` and 3 `# hadolint ignore=` directives:

```
detected languages: {'Python': 3}
pass 100.0  '0 suppressions / 3 LOC (0.0/1k)'
```

This is advertised in three user-facing places that are now false:
- the `Attribute.description` at [`code_quality.py:2044`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L2040-L2052) — **ships in every generated report**
- [`docs/attributes.md:1508`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/docs/attributes.md#L1508) — "Supported languages: … **Terraform** … **Dockerfile** …"
- the tests

**The tests certify the broken behavior.** [`test:1324`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/tests/unit/test_assessors_code_quality.py#L1317-L1326) and [`:1344`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/tests/unit/test_assessors_code_quality.py#L1337-L1346) hand-construct `languages={"Terraform": 1}` / `{"Dockerfile": 1}` — a state the scanner cannot emit, at a count below `minimum_file_threshold = 3`. They pass, and prove nothing about production.

### 3. Markdown/YAML in the denominator inverts the metric

`Markdown` and `YAML` are first-class scanned languages. Both are near-universally present (threshold is only 3 files) and both contribute their **full line count** to `total_lines` while carrying essentially zero suppressions. The metric is documented as "per 1,000 lines of **source code**".

Synthetic Go repo at a genuine 100 suppressions/1k lines, plus 24k lines of Markdown docs:

| Scope | Result |
|---|---|
| As shipped (Go + Markdown) | `pass`, **score 100.0** — 4.0/1k |
| Go source only | `fail`, **score 0.0** — 100.0/1k |

The score flips 0 → 100 purely because the project is documented — the exact inverse of the intended signal. On *this* repo the dilution is 3x (0.2/1k as shipped vs 0.6/1k Python-only).

Compounding it: [`code_quality.py:2158`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L2155-L2159) uses `len(text.splitlines())`, counting blank lines, whereas `LanguageDetector.count_total_lines()` deliberately excludes them — a further 15–25% denominator inflation. `splitlines()` also splits on `\v \f \x1c \x1d \x1e \x85    `; a file with 100 form-feed separators counted 302 lines against 202 real ones. Every one of these errors pushes toward a passing score.

### 4. Three silent-failure paths under-count the numerator while still reporting `pass`/100

All reproduced:

- **`lstat()` `OSError` → `continue`** ([`:2136-2139`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L2136-L2139)). A **sparse checkout** lists index entries that aren't on disk; every one is silently discarded. A repo at a true 714/1k (maximal fail) reported `pass, score 100.0, 0 suppressions / 1 LOC`.
- **The 1 MB read cap** ([`:2152`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L2150-L2154)). A 1.79 MiB `big.go` with 2000 `//nolint` past the cap → `pass, 0 suppressions / 17,139 LOC` (true: 32,001 lines). Large files are precisely where generated/vendored suppressions cluster.
- **`_count_file_suppressions` returns `(0,0)` on `OSError`** — byte-identical to "a clean empty file". A mode-000 file holding 50 `# noqa` vanished entirely: absent from the count, the denominator, *and* the top-offenders list.

Note that [`test_file_read_is_byte_bounded`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/tests/unit/test_assessors_code_quality.py#L1437-L1446) **asserts this silence is correct** — it hides 20 real suppressions past a patched cap and asserts `status == "pass"` and `"Total suppressions: 0"`. It should assert that truncation is *disclosed*.

For a density metric, a swallowed read is worse than a crash: it produces a falsely **good** certification, and nobody reads a passing finding's evidence.

### 5. `AGENTS.md` Guideline 7 — docs not synced for the two defunded weights

- [`docs/attributes.md:1153`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/docs/attributes.md#L1153) still documents `cyclomatic_complexity` as **Weight: 2%** (now `0.01`)
- [`docs/attributes.md:1330`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/docs/attributes.md#L1330) still documents `architectural_boundaries` as **Weight: 2%** (now `0.01`)

The diff does touch `docs/attributes.md`, but misses the two attributes it defunded.

---

## 🟡 Important

| # | Finding | Location |
|---|---|---|
| 6 | **All three threshold comparisons survive mutation.** `<=`→`<` at `:2210` and `:2217`, `>=`→`>` at `:2212` each leave 63/63 green — no test sits on a boundary. A flip at `:2217` silently converts every threshold-exact repo from pass to fail. Confirmed current behavior: density exactly 5.0 → pass/100, exactly 15.0 → fail/0. | `code_quality.py:2210-2217` |
| 7 | `.mjs`, `.cjs`, `.pyi`, `.pyx`, `.zsh` **trigger language detection but are never scanned** — `_LANG_EXTENSIONS` is a hand-maintained second copy of `EXTENSION_MAP` that has already drifted. An ESM-only JS repo scores 100 regardless of `eslint-disable` count. Also `.mdx` is in the assessor table but not the detector. | `code_quality.py:1972-1984` |
| 8 | **`Config.assessor_options` is dead public schema** — zero readers in `src/`, `tests/`, `docs/`, and absent from both example configs. It punches a hole in `extra="forbid"`: config written under it validates cleanly and does nothing, which is strictly worse than the loud typo it would otherwise be. It also serializes into every report via `assessment.py:128` while `CURRENT_SCHEMA_VERSION` stays `1.0.0`. Config schemas are one-way doors. | `models/config.py:159` |
| 9 | Total measurement failure reports as `not_applicable` ("No source files found for supported languages") — on a repo the scanner just reported as 100% Go. Contradicts the scan and misdirects debugging. `is_applicable()` has already established the language *is* present. | `code_quality.py:2194` |
| 10 | The bare `except Exception` around `git ls-files` hardcodes "missing git, not a git repo, or timeout" — but the **likeliest real trigger is `SubprocessSecurityError`** from `safe_subprocess_run`'s 10 MB stdout cap, which `git ls-files` clears on a ~200k-path monorepo. The user is told to install git, which is installed. `safe_subprocess_run_stream` already exists for exactly this. ("not a git repo" is near-unreachable — `Repository.__post_init__` rejects those already.) | `code_quality.py:2124` |
| 11 | `git ls-files` is re-invoked **once per detected language** for identical output — 4 spawns on this repo, 9 on a polyglot monorepo, each with a 30s timeout. Hoist it out of the loop. | `code_quality.py:2116` |
| 12 | `Attribute.criteria` hardcodes `_SUPPRESSION_DEFAULTS` while `Finding.threshold` honors config → a repo setting `pass_per_kloc: 10.0` gets one report containing both "≤5.0" and "≤10.0/1k". | `code_quality.py:2056` |

---

## 🔵 Suggestions

- The class docstring's *"Other languages always include all files"* ([`:2029`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/src/agentready/assessors/code_quality.py#L2029-L2031)) is false — `_TEST_DIR_FRAGMENTS` is checked *before* the per-language branch, so `tests/foo.sh` and `spec/values.yaml` **are** excluded. Only the filename-suffix heuristics are language-specific.
- [`docs/attributes.md:1504`](https://github.com/ambient-code/agentready/blob/6b3581f4f470a5f7feab345ccc98684598468656/docs/attributes.md#L1504) breaks house style — no `**ID**`/`**Tier**`/`**Weight**`/`**Category**`/`**Status**` block, no `#### Definition` / `Why It Matters` / `Measurable Criteria`, no `**Tools**` / `**Citations**`. Compare `### Threat Model` immediately above. The inline-`(id, weight)` heading format is the convention for the *bullet lists*, not for full `###` sections.
- The example-config comment `exclude_tests: false  # set true to exclude test files from LOC count` describes the wrong effect — the flag skips files *entirely* (numerator **and** denominator). Opposite-signed from what someone tuning it would predict. The `LintSuppressionOptions` docstring gets this right; the YAML comments contradict it.
- `_MAX_SUPPRESSION_FILE_BYTES = 1_000_000  # 1 MiB` — that's 1 MB. Same in `docs/attributes.md:1510`. Either say "1 MB" or use `1 << 20`.
- `default-weights.yaml:12` and `:72` still say Tier 3 has **8 attributes**; it's 9. (Line 3 was correctly bumped to 30.)
- The changelog block labels this `v2.5.0`, but `pyproject.toml` is at **2.50.0** and `lint_config_coverage` — labeled `v2.4.0` — actually shipped in 2.50.0. The version numbering in that block has drifted from real releases; consider keying entries by branch/issue instead.
- `pass_per_kloc` uses `gt=0`, forbidding `0.0` — the legitimate strictest policy ("no suppressions tolerated"). `ge=0` on pass, `gt=0` on fail.
- `_get_options` unpacks the validated model into `tuple[float, float, bool]` — two positionally-adjacent floats of identical type, reintroducing the transposition hazard the validator exists to prevent. Return the `LintSuppressionOptions` instead.
- Consider `frozen=True` on `LintSuppressionOptions`: pydantic v2 doesn't re-run `mode="after"` validators on assignment, and `_SUPPRESSION_DEFAULTS` is a process-wide shared mutable instance.

---

## ✅ Strengths

- **The security posture is genuinely strong.** git-only discovery with a deliberate no-fallback policy (well-commented at `:2113` — exactly the kind of "why" comment that stops a future maintainer from helpfully adding the fallback back), `lstat` + `stat.S_ISREG` symlink rejection, bounded reads. Mutation probes confirmed the symlink and gitignore tests each catch their regression, and the symlink test's use of a gitignored target to isolate the behavior is well-judged.
- `_GitInventoryUnavailable` is the best-designed type in the diff — it cleanly separates "cannot measure" (`skipped`) from "measured zero" (`not_applicable`) from "bug" (`error`), and makes the no-silent-fallback decision greppable.
- Weight arithmetic is correct: 30 attributes summing to exactly `1.0`, tiers 58/27/13/2, and the new attribute is properly funded by reductions elsewhere.
- `black`, `isort`, `ruff check` all clean. Line coverage of the new code is ~97%.
- Assessor registration follows the `AGENTS.md` 4-step checklist correctly.
- The custom two-threshold interpolation instead of `calculate_proportional_score()` is a defensible deviation — the helper models a single threshold and can't express a pass/fail band.

---

## Recommended sequence

1. **Fix the red test** (#1) — unblocks CI.
2. **Decide Terraform/Dockerfile** (#2): add `.tf` + a filename rule to `EXTENSION_MAP`, or strip them from the pattern tables, the `Attribute.description`, the docs, and the two tests. Then add the guard that would have caught this pre-merge:
   ```python
   def test_suppression_tables_agree_with_detector():
       assert set(_SUPPRESSION_PATTERNS) == set(_LANG_EXTENSIONS)
       assert set(_SUPPRESSION_PATTERNS) <= set(LanguageDetector.EXTENSION_MAP.values())
   ```
3. **Fix the denominator** (#3): exclude Markdown/YAML from LOC (or report per-language), and stop counting blank lines. This is the change that most affects what the assessor actually measures.
4. **Make silent skips visible** (#4): thread a scan-health tally (`files_missing`, `files_unreadable`, `files_truncated`) through the walk, and emit it in `evidence` unconditionally — on pass as well as fail. If the un-scanned ratio exceeds ~10%, downgrade to `skipped` rather than reporting a density at all. A density computed over an unknown fraction of the codebase isn't a measurement.
5. **Derive `_LANG_EXTENSIONS` from `EXTENSION_MAP`** (#7) — structurally eliminates this entire drift class.
6. **Drop `assessor_options`** (#8) — re-add it in the PR that first needs it, with a reader, a precedence rule, and a test.
7. Sync the two weight docs (#5), add boundary tests (#6), re-run.

---

The security-critical paths here were clearly thought about carefully, and the no-fallback discipline around `git ls-files` is a good instinct. The gap is that the **measurement** paths weren't held to the same standard: the language table was written without checking it against the detector, and several tests confirm the implementation's assumptions rather than the behavior users will get.

<sub>Posted by an AI agent (Claude Opus 4.6 via Claude Code, `/pr-review-toolkit:review-pr`). Findings are reproducible — please verify before acting, and correct anything that's wrong.</sub>

**Comment by msu8** [comment]
Hmm, a very suspicious CI fail with 7k lines

**Comment by kami619** [comment]
> Hmm, a very suspicious CI fail with 7k lines

@msu8 ignore that, it was from a bad merge. I overlooked this CI failure from a different PR., I shall post a fix soon and you can rebase on your end.

**Comment by msu8** [comment]
Got it. Anyway, I think I addressed both your review and the coderabbit stuff in the last two commits.
